### ISSUE TITLE
Appending instead of reloading offsets of XDR formats

### ISSUE THREAD

**Issue by yuxuanzhuang** [MDAnalysis/mdanalysis#3225]
## Is your feature request related to a problem? ##
It is slow to access a large XDR-based trajectory (~50 GB), e.g. XTC file, mainly because an offset file has to be built for random accessing every frame (benchmark needed). When one tries to run some preliminary analysis on a running trajectory, the offset needs to be reconstructed every time the trajectory is updated.

## Describe the solution you'd like ##
I am wondering if it is possible to append to the existing offset instead of rebuilding one assuming it will reduce the time. (ofc after checking the offset matches the trajectory)

## Describe alternatives you've considered ##
Optimize the offset construction time.

**Comment by IAlibay** [comment]
Going to tag @hmacdope here in the hope that there might have been some benchmarks/lessons learned with the TNG work.

**Comment by richardjgowers** [comment]
I'm pretty sure that the act of checking that the new trajectory matches the old one is probably identically expensive to constructing the offsets.  I think the real solution to this will be TNG.  You could probably write something hacky which builds offsets starting at a given location, but it's not really something stable enough that we'd want to include and support.

**Comment by hmacdope** [comment]
TNG can append any block to any trajectory frame set (block of frames) and access is done by simple in-file pointers. I think this means you don't have to traverse the entire file every time, just chase the right pointers to your block (I think).

 Give it a try and see if it makes a difference!

I think you can convert with `gmx trjconv -f [].xtc -s [].tpr -o [].tng`

**Comment by hmacdope** [comment]
As a side note from my memory, even `gmx-check` and `gmx-dump` are not that fast for xtc files.

**Comment by yuxuanzhuang** [comment]
> it's not really something stable enough that we'd want to include and support.

Thanks! I understand it might be buggy and hard to maintain but it's just pretty annoying to reload the offset every time (which is bad for achieving fast in-situ analyses in the supercomputer). Guess I will look forward to the TNG era:)

**Comment by yuxuanzhuang** [comment]
> As a side note from my memory, even `gmx-check` and `gmx-dump` are not that fast for xtc files.

Yeah, I think it is pretty slow to iterate over the whole XTC file even when I only want to get info for the last frame; in that case, defining `-b lastframe` can significantly increase the speed for  `gmx trjconv -dump lastframe` for some reason I am not so sure about.

**Comment by orbeckst** [comment]
In principle one could implement heuristics that allow you to seek into the vicinity of a xtc frame and then start doing a binary search. I assume that’s what the gmx tools do. But I don’t know if there’s anyone willing to write such code for MDA, especially code that does not on occasion produce wrong results. Part of such a patch would be an analysis of its robustness. 

Ultimately using a better file format seems to be the better choice. 

(On that note: where are we with a proper TNGReader for MDA, @hmacdope ?)

**Comment by hmacdope** [comment]
@orbeckst aside from topology information pytng is fully functional and stable at `0.2.2`, I'm just not exactly sure how we want to package it into MDA itself. 

Do we want to submodule? Copy source code? This is the part I wasn't sure about. :)

**Comment by appassionate** [comment]
Hi,  I have an idea to  accelerate offset finding process based on those facts.
### some  coding findings
- f.seek() with f.readline() will return None when seek arrive at "no data region" in a filestream (maybe some advanced class in MDA)
- thanks to the pickleable features in MDA2, we can use filestream freely in parallel way with multiprocessing or any other tools for python parallel.
- almost all file format will obey a periodic form, so we can locate a complete structure by simple operating.

### some things should do 
- find the maximum "seek index" in filestream in some way.
- split this seek idx range as tasks and find offset index in parallel
- locate the nearest frame heading in a any offered "seek index", and presented as a offset value, which might using a regular expression to match for every format of files.

Actually, I have an demo for it . It is a simple try to process ".XYZ" file with regular expression matching. it seem works, but have limited improvement which will relate to CPU numbers and I/O maximum performance?  I guess..    : )  if you are interested in it, I will post some code in this issue... 

hope mda will update XYZReader with reloading offsets functions... haha

**Comment by orbeckst** [comment]
Just re-reading this issue.

I am in favor of closing with "wont fix" because this seems potentially buggy and hard to maintain.

If someone demonstrates via code that it's robust and reliable then I'll change my mind.

If I don't hear anything substantial over the next week I'll close with "wont fix".

**Comment by orbeckst** [comment]
@appassionate if you're interested in something related to XYZ please open a new issue so that discussions can remain focused. Thank you!

**Comment by orbeckst** [comment]
Btw, @yuxuanzhuang I share your frustration in having to rebuild offsets for analyzing running simulations.

**Comment by yuxuanzhuang** [comment]
Sounds reasonable to me. A better format does seem to be a better solution:)
